internal/events/loop.go

```diff
@@ -5,158 +5,94 @@ import (
+// FIXME(pleshakov): better document the side effects and how to prevent and mitigate them.
+// So when the EventLoop have 100 saved events, it is better to process them at once rather than one by one.
 type EventLoop struct {
-	cfg EventLoopConfig
+	eventCh <-chan interface{}
+	logger  logr.Logger
+	handler EventHandler
 }
 
 // NewEventLoop creates a new EventLoop.
-func NewEventLoop(cfg EventLoopConfig) *EventLoop {
-	return &EventLoop{cfg: cfg}
+func NewEventLoop(eventCh <-chan interface{}, logger logr.Logger, handler EventHandler) *EventLoop {
+	return &EventLoop{
+		eventCh: eventCh,
+		logger:  logger,
+		handler: handler,
+	}
 }
 
 // Start starts the EventLoop.
-// The method will block until the EventLoop stops:
-// - if it stops because of an error, the Start will return the error.
-// - if it stops normally, the Start will return nil.
+// This method will block until the EventLoop stops, which will happen after the ctx is closed.
+//
+// FIXME(pleshakov). Ensure that when the Gateway starts, the first time it generates configuration for NGINX,
+// it has a complete view of the cluster resources. For example, when the Gateway processes a Gateway resource
+// with a listener with TLS termination enabled (the listener references a TLS Secret), the Gateway knows about the secret.
+// This way the Gateway will not produce any incomplete transient configuration at the start.
 func (el *EventLoop) Start(ctx context.Context) error {
+	// The current batch.
+	var batch EventBatch
+	// handling tells if any batch is currently being handled.
+	var handling bool
+	// handlingDone is used to signal the completion of handling a batch.
+	handlingDone := make(chan struct{})
+
+	handle := func(ctx context.Context, batch EventBatch) {
+		el.handler.HandleEventBatch(ctx, batch)
+		handlingDone <- struct{}{}
+	}
+
+	// Note: at any point of time, no more than one batch is currently being handled.
+
+	// The event loop
 	for {
 		select {
 		case <-ctx.Done():
-			// although we always return nil, Start must return it to satisfy
-			// "sigs.k8s.io/controller-runtime/pkg/manager".Runnable
+			// Wait for the completion if a batch is being handled.
+			if handling {
+				<-handlingDone
+			}
 			return nil
-		case e := <-el.cfg.EventCh:
-			el.handleEvent(ctx, e)
+		case e := <-el.eventCh:
+			// Add the event to the current batch.
+			batch = append(batch, e)
+
+			// FIXME(pleshakov): Log more details about the event like resource GVK and ns/name.
+			el.logger.Info(
+				"added an event to the current batch",
+				"type", fmt.Sprintf("%T", e),
+				"total", len(batch),
+			)
+
+			// Handle the current batch if no batch is being handled.
+			if !handling {
+				go handle(ctx, batch)
+				batch = make([]interface{}, 0)
+				handling = true
+			}
+		case <-handlingDone:
+			handling = false
+
+			// Handle the current batch if it has at least one event.
+			if len(batch) > 0 {
+				go handle(ctx, batch)
```

Comment:
Each time we create the goroutine I think a Debug log of what we're processing is useful. Here and 83.

Reply:
Added two log statements in the `handle()` code. Also, removed a few and changed in `handler.go`
I used `Info()` for logging as the existing similar ones use `Info()`.